invenio_sequencegenerator/models.py

```diff
@@ -0,0 +1,202 @@
+            seq = db.session.query(cls).filter_by(name=id).first() or \
+                  db.session.query(cls).filter_by(template=id).first()
+            if seq is None:
+                raise NoSuchSequence()
+
+            try:
+                template = seq.template.format(**kwargs)
+            except KeyError:
+                raise InvalidCall('Improper arguments for string formatting')
+
+            if seq.parent:
+                # Check if parent's report number exists in PID store
+                try:
+                    get_distribution('invenio-pidstore')
+                    # from invenio_pidstore.errors import PIDDoesNotExistError
+                    from invenio_pidstore.models import PersistentIdentifier
+
+                    try:
+                        PersistentIdentifier.query.filter_by(
+                            pid_value=template
+                        ).one()
+                    except NoResultFound:
+                        raise InvalidCall('Parent PID does not exist')
+                except DistributionNotFound:
+                    pass
+
+                return ReportNumber.next_or_create(template)
+            else:
+                return ReportNumber.next_or_create(template)
+
+
+class ReportNumber(db.Model):
+    """Stores generated report numbers."""
+
+    __tablename__ = 'sequencegenerator_reportnumber'
+
+    template = db.Column(db.String, nullable=False,
+                         primary_key=True, index=True)
+
+    counter = db.Column(db.Integer, nullable=False)
+
+    @classmethod
+    def create(cls, template, counter=1):
+        """Create a new report number."""
+        with db.session.begin_nested():
+            rec = cls(
+                template=template,
+                counter=counter
+            )
+            db.session.add(rec)
+        return rec
+
+    @classmethod
+    def next_or_create(cls, template):
+        """Get the next report number for the given template."""
+        with db.session.begin_nested():
+            rec = db.session.query(cls).filter_by(template=template).first()
+
+            if rec is None:
+                rec = ReportNumber.create(template)
+            else:
+                rec.counter += 1
+
+        return rec.template.format(counter=rec.counter)
+
+
+#
+# User-level API
+#
+def new_sequence(name, template):
+    """Create a new sequence."""
+    Sequence.create(name, template)
+
+
+def get_next(name_or_template, **kwargs):
+    """Return the next available report number for the given template."""
+    next = Sequence.next(name_or_template, **kwargs).encode('ascii', 'ignore')
+
+    return next
+
+
+#
+# Template utilities
+#
+def extract_placeholders(template):
+    """Extract the template's placeholder names."""
+    return re.findall(r'{(.*?)}', template)
+
+
+def double_counter(template):
+    """Double brackets around the 'counter' for 2-step formatting."""
+    return re.sub(counter_regex, r'{\1}', template)
+
+
+#
```

Comment:
Can you move exceptions to `errors.py`?

--
Example: https://github.com/inveniosoftware/invenio-pidstore/blob/master/invenio_pidstore/errors.py